Added port control indicator to advanced settings

kennysong opened the pull request:
After a user logs into their first social network, uProxy now probes the router for port control support, and displays an indicator in the advanced settings page. There's also an option for manually re-running the probe next to the indicator. An FAQ entry was also added about port control.

<!-- Reviewable:start -->
[<img src="https://reviewable.io/review_button.png" height=40 alt="Review on Reviewable"/>](https://reviewable.io/reviews/uproxy/uproxy/1754)
<!-- Reviewable:end -->

kennysong commented:
@jpevarnek @bemasc

bemasc commented on src/generic_ui/faq.html:
```diff
@@ -126,6 +127,13 @@ <h1 class="uproxy-title"><span class="light">u</span>Proxy FAQ</h1>
 
+          <section class="faq" id="whatIsPortControl">
+            <h2>What does "port control" mean?</h2>
+            <p>If your router has port control enabled (UPnP, NAT-PMP, or PCP), uProxy can talk to your router to specifically allow incoming connections from other uProxy users. If just one side of a uProxy connection has port control, sharing/getting internet access will almost always be successful. (However, successful connections can be made without port control.) To learn how to enable port  ...
```
s/just one side/either side/

bemasc commented on src/generic_core/uproxy_core.ts:
```diff
@@ -100,6 +92,7 @@ export class uProxyCore implements uproxy_core_api.CoreApi {
   }
 
   private pendingNetworks_ :{[name :string] :social.Network} = {};
+  private portControlSupport = false;
```
I think this should be a tri-state (true, false, or pending), with corresponding UI.  (Also, private fields should have a trailing underscore).

kennysong commented:
<sup>**[src/generic_core/uproxy_core.ts, line 95 \[r1\]](https://reviewable.io:443/reviews/uproxy/uproxy/1754#-Jv-teM_Z_bfd_yPjgRD-r1-95)** ([raw file](https://github.com/uproxy/uproxy/blob/1da96e13d9767a5429a1d2b62d8d174dea49aa75/src/generic_core/uproxy_core.ts#L95)):</sup>
Should the tri-state be a string or enum?

---


Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/uproxy/uproxy/1754)
<!-- Sent from Reviewable.io -->

bemasc commented:
Review status: 0 of 11 files reviewed at latest revision, 2 unresolved discussions, all commit checks successful.

---

<sup>**[src/generic_core/uproxy_core.ts, line 95 \[r1\]](https://reviewable.io:443/reviews/uproxy/uproxy/1754#-Jv-teM_Z_bfd_yPjgRD-r1-95)** ([raw file](https://github.com/uproxy/uproxy/blob/1da96e13d9767a5429a1d2b62d8d174dea49aa75/src/generic_core/uproxy_core.ts#L95)):</sup>
I recommend an enum

---


Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/uproxy/uproxy/1754)
<!-- Sent from Reviewable.io -->

jpevarnek commented:
Since I don't think you have ever done a code review with me before, a few quick notes:

* You should interpret every statement I make as me suggesting there should be discussion around a topic.  If I make a statement that sounds even slightly off to you, don't make the change, just reply to the comment.  If anything sounds like I'm telling you to definitely make a change, it's just that I forgot to reword it...
* If I look at an html file, I've probably just glanced at the rules and made sure that there is nothing that jumps out there, I have not looked at how anything looks.


General comments on the PR:
Everything here looks great overall, definitely a good thing to get this into uProxy.  I think it would be nice to make sure that this design is easily expandable to also show something if we are using the port-control functionality (assuming we want to show it in a similar way), so if you could write a sentence or two on how you would extend this to that situation, that would be amazing!

The one thing that I'm not really sure about (overall design wise) is that we have so many different places where we are either setting or computing the port-control levels.  It seems to me that having the single boolean value on both the UI and server side would be a simpler design and, assuming that was done, I think you could get a simpler design by switching the semantics of the command to just instruct the core to perform the refresh and then rely on the core to, at some later point (iff the value changed), just send back the update.  That way, you would only have to handle changes at one point on the UI side (aside from initial state) and only have to handle the calculation at one point on the core side.  Thoughts?

---

Reviewed 11 of 11 files at r1.
Review status: 4 of 11 files reviewed at latest revision, 10 unresolved discussions, some commit checks pending.

---

<sup>**[src/generic_core/uproxy_core.ts, line 400 \[r1\]](https://reviewable.io:443/reviews/uproxy/uproxy/1754#-Jv0EYyJfwWyBEER2uK6)** ([raw file](https://github.com/uproxy/uproxy/blob/1da96e13d9767a5429a1d2b62d8d174dea49aa75/src/generic_core/uproxy_core.ts#L400)):</sup>
Is there an interface you can use for this?  In general, we try to avoid using `any` as a type.

---

<sup>**[src/generic_core/uproxy_core.ts, line 413 \[r1\]](https://reviewable.io:443/reviews/uproxy/uproxy/1754#-Jv0EHgltdNr2NbxxOa0)** ([raw file](https://github.com/uproxy/uproxy/blob/1da96e13d9767a5429a1d2b62d8d174dea49aa75/src/generic_core/uproxy_core.ts#L413)):</sup>
My general intuition here would be that we should have a single refresh function and then call it only if it was the first network.

If not, we should at least have this start with a verb (e.g. `refreshPortControlIfSingleLogin`)

---

<sup>**[src/generic_core/uproxy_core.ts, line 417 \[r1\]](https://reviewable.io:443/reviews/uproxy/uproxy/1754#-Jv0UiNO04KPUwp6o2rS)** ([raw file](https://github.com/uproxy/uproxy/blob/1da96e13d9767a5429a1d2b62d8d174dea49aa75/src/generic_core/uproxy_core.ts#L417)):</sup>
This form of writing is a bit odd and hard to understand.  Points that are a bit confusing:

* Separating `networkNames` from `social_networks.networks` makes it a bit unintuitive that you are simply indexing over `social_network.networks` (recommendation: use `for (x in object)` syntax)
* `Object.keys(x).length` takes a tiny bit of time to understand (recommendation: use `_.size`)

  My personal preference would be something closer to:

`numNetworks = _.reduce(social_networks.networks, (memo, networks) => { return memo + (_.size(networks) ? 1 : 0); }, 0);`

(except, you know, not on one line, that's terrible and hard to read) or

```
for (name in social_networks.networks) {
  if (_.size(social_networks.networks[name]) > 0) {
    numNetworks++;
  }
}
```

if you want to be more verbose.

---

<sup>**[src/generic_core/uproxy_core.ts, line 426 \[r1\]](https://reviewable.io:443/reviews/uproxy/uproxy/1754#-Jv0X55_YCPmOQ91a1Za)** ([raw file](https://github.com/uproxy/uproxy/blob/1da96e13d9767a5429a1d2b62d8d174dea49aa75/src/generic_core/uproxy_core.ts#L426)):</sup>
It seems like this could be easier done with `ui.update(uproxy_core_api.Update.LOGIN_PORT_CONTROL_STATUS, this.portControlSupport)`

Anyways, in general, I'd try to avoid the duplication of having the separate `ui.update` calls here with the only difference being constant parameters.  Instead, I'd try to calculate the parameter beforehand.

---

<sup>**[src/generic_ui/faq.html, line 130 \[r1\]](https://reviewable.io:443/reviews/uproxy/uproxy/1754#-Jv0XfgIH3MuST66euC-)** ([raw file](https://github.com/uproxy/uproxy/blob/1da96e13d9767a5429a1d2b62d8d174dea49aa75/src/generic_ui/faq.html#L130)):</sup>
Make sure to either add this to the website once the code goes out or bug Lucy or me to do so.

---

<sup>**[src/generic_ui/polymer/advanced-settings.html, line 91 \[r1\]](https://reviewable.io:443/reviews/uproxy/uproxy/1754#-Jv0ZPJg5tbX_knoohv3)** ([raw file](https://github.com/uproxy/uproxy/blob/1da96e13d9767a5429a1d2b62d8d174dea49aa75/src/generic_ui/polymer/advanced-settings.html#L91)):</sup>
I'd simplify this rule to `#portControlBox core-icon` (since these are the only two).  If not that, add a class to them.  It seems very unlikely what you actually want to say here is "Apply this styling to any core-icon within an element with the id `portControlBox` that happens to be either the help or refresh icon"

---

<sup>**[src/generic_ui/polymer/advanced-settings.html, line 106 \[r1\]](https://reviewable.io:443/reviews/uproxy/uproxy/1754#-Jv0qO8j9ont8KPRWQ4u)** ([raw file](https://github.com/uproxy/uproxy/blob/1da96e13d9767a5429a1d2b62d8d174dea49aa75/src/generic_ui/polymer/advanced-settings.html#L106)):</sup>
I'm not positive, but my basic instinct looking at this is that you are looking for the rule `#portControlBox>*`.  What are you trying to accomplish semantically here?

---

<sup>**[src/generic_ui/polymer/advanced-settings.html, line 110 \[r1\]](https://reviewable.io:443/reviews/uproxy/uproxy/1754#-Jv0qeNdv_yv8pTh-0eG)** ([raw file](https://github.com/uproxy/uproxy/blob/1da96e13d9767a5429a1d2b62d8d174dea49aa75/src/generic_ui/polymer/advanced-settings.html#L110)):</sup>
I know it makes no difference result-wise, but I think you might want this to semantically be `#portControlBox .label`.  Justification: You are looking for all text that is a label, not all spans.

---


Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/uproxy/uproxy/1754)
<!-- Sent from Reviewable.io -->

kennysong commented:
Review status: 4 of 11 files reviewed at latest revision, 10 unresolved discussions, some commit checks pending.

---

<sup>**[src/generic_core/uproxy_core.ts, line 95 \[r1\]](https://reviewable.io:443/reviews/uproxy/uproxy/1754#-Jv-teM_Z_bfd_yPjgRD-r1-95)** ([raw file](https://github.com/uproxy/uproxy/blob/1da96e13d9767a5429a1d2b62d8d174dea49aa75/src/generic_core/uproxy_core.ts#L95)):</sup>
Done.

---

<sup>**[src/generic_ui/faq.html, line 132 \[r1\]](https://reviewable.io:443/reviews/uproxy/uproxy/1754#-Jv-teMa2xIMIpc-PM4L-r1-132)** ([raw file](https://github.com/uproxy/uproxy/blob/1da96e13d9767a5429a1d2b62d8d174dea49aa75/src/generic_ui/faq.html#L132)):</sup>
Done.

---


Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/uproxy/uproxy/1754)
<!-- Sent from Reviewable.io -->

jpevarnek commented:
Reviewed 7 of 7 files at r2.
Review status: all files reviewed at latest revision, 10 unresolved discussions, some commit checks pending.

---

<sup>**[src/generic_core/uproxy_core.ts, line 426 \[r1\]](https://reviewable.io:443/reviews/uproxy/uproxy/1754#-Jv0X55_YCPmOQ91a1Za)** ([raw file](https://github.com/uproxy/uproxy/blob/1da96e13d9767a5429a1d2b62d8d174dea49aa75/src/generic_core/uproxy_core.ts#L426)):</sup>
Update based on new commit: I'd still say just have a single `ui.Update` and then have the value as a variable.

---


Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/uproxy/uproxy/1754)
<!-- Sent from Reviewable.io -->

bemasc commented:
Review status: all files reviewed at latest revision, 10 unresolved discussions, all commit checks successful.

---

<sup>**[src/generic_core/uproxy_core.ts, line 416 \[r1\]](https://reviewable.io:443/reviews/uproxy/uproxy/1754#-Jv0EHgltdNr2NbxxOa0)** ([raw file](https://github.com/uproxy/uproxy/blob/1da96e13d9767a5429a1d2b62d8d174dea49aa75/src/generic_core/uproxy_core.ts#L416)):</sup>
If this UI moves to the Advanced Settings, then I don't think we need to run this at login.  It can just run whenever you open Advanced Settings.

---


Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/uproxy/uproxy/1754)
<!-- Sent from Reviewable.io -->

kennysong commented:
If we wanted to display another state, say "Port control is active", then the enum could be expanded to include `ACTIVE`, which would be set in  `configureProbeConnection_()` in `churn.ts` where we open a port for a new connection. After the connection is successful, `core.portControlSupport_` can be updated and an `ui.update` pushed to the UI. Same when a connection is closed.

I also simplified the way the probe data floats between the core and UI. We no longer touch `ui.portControlSupport` in `advanced-settings.ts`. `ui.portControlSupport` is only set via a `ui.update()` sent from the core, and handled solely inside `ui.ts`. This also means that we are calling `refreshPortControlSupport()` and ignoring its return Promise, but we can access that (individual protocol) data easily if needed later on.

---

Review status: 5 of 11 files reviewed at latest revision, 10 unresolved discussions, some commit checks pending.

---

<sup>**[src/generic_core/uproxy_core.ts, line 400 \[r1\]](https://reviewable.io:443/reviews/uproxy/uproxy/1754#-Jv0EYyJfwWyBEER2uK6)** ([raw file](https://github.com/uproxy/uproxy/blob/1da96e13d9767a5429a1d2b62d8d174dea49aa75/src/generic_core/uproxy_core.ts#L400)):</sup>
Done.

---

<sup>**[src/generic_core/uproxy_core.ts, line 413 \[r1\]](https://reviewable.io:443/reviews/uproxy/uproxy/1754#-Jv0EHgltdNr2NbxxOa0)** ([raw file](https://github.com/uproxy/uproxy/blob/1da96e13d9767a5429a1d2b62d8d174dea49aa75/src/generic_core/uproxy_core.ts#L413)):</sup>
Done. There's now only one function that's called when the advanced settings pane is opened, and when the refresh button is clicked.

I also removed `portControlSupport` from the `InitialState` object, since we no longer need to persist the value if the uProxy window is closed (it's recalculated when advanced settings is opened).

---

<sup>**[src/generic_core/uproxy_core.ts, line 426 \[r1\]](https://reviewable.io:443/reviews/uproxy/uproxy/1754#-Jv0X55_YCPmOQ91a1Za)** ([raw file](https://github.com/uproxy/uproxy/blob/1da96e13d9767a5429a1d2b62d8d174dea49aa75/src/generic_core/uproxy_core.ts#L426)):</sup>
Done.

---

<sup>**[src/generic_ui/polymer/advanced-settings.html, line 91 \[r1\]](https://reviewable.io:443/reviews/uproxy/uproxy/1754#-Jv0ZPJg5tbX_knoohv3)** ([raw file](https://github.com/uproxy/uproxy/blob/1da96e13d9767a5429a1d2b62d8d174dea49aa75/src/generic_ui/polymer/advanced-settings.html#L91)):</sup>
Done.

---

<sup>**[src/generic_ui/polymer/advanced-settings.html, line 106 \[r1\]](https://reviewable.io:443/reviews/uproxy/uproxy/1754#-Jv0qO8j9ont8KPRWQ4u)** ([raw file](https://github.com/uproxy/uproxy/blob/1da96e13d9767a5429a1d2b62d8d174dea49aa75/src/generic_ui/polymer/advanced-settings.html#L106)):</sup>
Removed these CSS rules – they weren't actually necessary.

---

<sup>**[src/generic_ui/polymer/advanced-settings.html, line 110 \[r1\]](https://reviewable.io:443/reviews/uproxy/uproxy/1754#-Jv0qeNdv_yv8pTh-0eG)** ([raw file](https://github.com/uproxy/uproxy/blob/1da96e13d9767a5429a1d2b62d8d174dea49aa75/src/generic_ui/polymer/advanced-settings.html#L110)):</sup>
Done.

---


Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/uproxy/uproxy/1754)
<!-- Sent from Reviewable.io -->

kennysong commented:
<sup>**[src/generic_core/uproxy_core.ts, line 420 \[r1\]](https://reviewable.io:443/reviews/uproxy/uproxy/1754#-Jv0UiNO04KPUwp6o2rS)** ([raw file](https://github.com/uproxy/uproxy/blob/1da96e13d9767a5429a1d2b62d8d174dea49aa75/src/generic_core/uproxy_core.ts#L420)):</sup>
Thanks! This function is completely deleted in the newest version, so this code doesn't exist anymore. 

However, I think that for `for...in` loop for an object will go through all the properties in its prototype chain as well, so you would need a 
`if (social_networks.networks.hasOwnProperty(name))` check. (?)

---


Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/uproxy/uproxy/1754)
<!-- Sent from Reviewable.io -->

jpevarnek commented:
Reviewed 3 of 6 files at r3.
Review status: 8 of 11 files reviewed at latest revision, 8 unresolved discussions, all commit checks successful.

---

<sup>**[src/generic_core/uproxy_core.ts, line 97 \[r3\]](https://reviewable.io:443/reviews/uproxy/uproxy/1754#-JvEvdSr_R_0mEBysMww)** ([raw file](https://github.com/uproxy/uproxy/blob/b8b2c4fd104b4ec472b73b8da6f8457b59020486/src/generic_core/uproxy_core.ts#L97)):</sup>
My general instinct would be that if we are still storing this in the core, we should send it in the initial state update.  Thoughts?

---

<sup>**[src/generic_core/uproxy_core.ts, line 418 \[r1\]](https://reviewable.io:443/reviews/uproxy/uproxy/1754#-Jv0UiNO04KPUwp6o2rS)** ([raw file](https://github.com/uproxy/uproxy/blob/1da96e13d9767a5429a1d2b62d8d174dea49aa75/src/generic_core/uproxy_core.ts#L418)):</sup>
Technically, yes, but I don't believe it would be necessary in this specific instance.

---

<sup>**[src/generic_ui/scripts/core_connector.ts, line 216 \[r3\]](https://reviewable.io:443/reviews/uproxy/uproxy/1754#-JvEuGsCNale_QRpaJ4-)** ([raw file](https://github.com/uproxy/uproxy/blob/b8b2c4fd104b4ec472b73b8da6f8457b59020486/src/generic_ui/scripts/core_connector.ts#L216)):</sup>
If we are no longer going to care about the return value here, I'd switch this to either a void promise command or a non-promise command.

---

<sup>**[src/interfaces/uproxy_core_api.ts, line 134 \[r3\]](https://reviewable.io:443/reviews/uproxy/uproxy/1754#-JvEvBWCj6RiNiqnoFCf)** ([raw file](https://github.com/uproxy/uproxy/blob/b8b2c4fd104b4ec472b73b8da6f8457b59020486/src/interfaces/uproxy_core_api.ts#L134)):</sup>
The name here sounds more like a command than an update (something that is also a problem with other messages).  I'd call it something closer to `PORT_CONTROL_STATUS_CHANGED`

Feel free to ignore this comment.

---


Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/uproxy/uproxy/1754)
<!-- Sent from Reviewable.io -->

jpevarnek commented:
Reviewed 3 of 6 files at r3.
Review status: all files reviewed at latest revision, 8 unresolved discussions, all commit checks successful.

---



Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/uproxy/uproxy/1754)
<!-- Sent from Reviewable.io -->

kennysong commented:
<sup>**[src/generic_core/uproxy_core.ts, line 97 \[r3\]](https://reviewable.io:443/reviews/uproxy/uproxy/1754#-JvEvdSr_R_0mEBysMww)** ([raw file](https://github.com/uproxy/uproxy/blob/b8b2c4fd104b4ec472b73b8da6f8457b59020486/src/generic_core/uproxy_core.ts#L97)):</sup>
We could, but I took this out of the initial state update since it's not necessary for this feature. Every time the advanced settings template is loaded, it will automatically recalculate this value in the core and send it to the UI.

---


Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/uproxy/uproxy/1754)
<!-- Sent from Reviewable.io -->

kennysong commented:
Review status: 7 of 11 files reviewed at latest revision, 7 unresolved discussions, some commit checks pending.

---

<sup>**[src/generic_ui/scripts/core_connector.ts, line 216 \[r3\]](https://reviewable.io:443/reviews/uproxy/uproxy/1754#-JvEuGsCNale_QRpaJ4-)** ([raw file](https://github.com/uproxy/uproxy/blob/b8b2c4fd104b4ec472b73b8da6f8457b59020486/src/generic_ui/scripts/core_connector.ts#L216)):</sup>
Done.

---

<sup>**[src/interfaces/uproxy_core_api.ts, line 134 \[r3\]](https://reviewable.io:443/reviews/uproxy/uproxy/1754#-JvEvBWCj6RiNiqnoFCf)** ([raw file](https://github.com/uproxy/uproxy/blob/b8b2c4fd104b4ec472b73b8da6f8457b59020486/src/interfaces/uproxy_core_api.ts#L134)):</sup>
Done. I changed it to `PORT_CONTROL_STATUS` (since the status doesn't have to change for an update to be pushed).

---


Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/uproxy/uproxy/1754)
<!-- Sent from Reviewable.io -->

kennysong commented:
<sup>**[src/generic_ui/faq.html, line 130 \[r1\]](https://reviewable.io:443/reviews/uproxy/uproxy/1754#-Jv0XfgIH3MuST66euC-)** ([raw file](https://github.com/uproxy/uproxy/blob/1da96e13d9767a5429a1d2b62d8d174dea49aa75/src/generic_ui/faq.html#L130)):</sup>
It's here: https://github.com/uProxy/uproxy-website/pull/82

---


Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/uproxy/uproxy/1754)
<!-- Sent from Reviewable.io -->

jpevarnek commented:
Reviewed 4 of 4 files at r4.
Review status: all files reviewed at latest revision, 6 unresolved discussions, all commit checks successful.

---

<sup>**[src/generic_core/uproxy_core.ts, line 97 \[r3\]](https://reviewable.io:443/reviews/uproxy/uproxy/1754#-JvEvdSr_R_0mEBysMww)** ([raw file](https://github.com/uproxy/uproxy/blob/b8b2c4fd104b4ec472b73b8da6f8457b59020486/src/generic_core/uproxy_core.ts#L97)):</sup>
Okay, is there any reason we are saving it here then?

If we are not treating it as a value within the core that the UI should be updated on, I'd actually reverse my previous recommendations and say that it should be sent to the UI only through the return value of the call and then stored in the Polymer state instead of in the UI at all.

---


Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/uproxy/uproxy/1754)
<!-- Sent from Reviewable.io -->

kennysong commented:
<sup>**[src/generic_core/uproxy_core.ts, line 97 \[r3\]](https://reviewable.io:443/reviews/uproxy/uproxy/1754#-JvEvdSr_R_0mEBysMww)** ([raw file](https://github.com/uproxy/uproxy/blob/b8b2c4fd104b4ec472b73b8da6f8457b59020486/src/generic_core/uproxy_core.ts#L97)):</sup>
Okay, for consistency and since this value might be used for other features later on (e.g. displaying a port control status in your friends list), I added it back into the core initial state update.

---


Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/uproxy/uproxy/1754)
<!-- Sent from Reviewable.io -->

jpevarnek commented:
:thumbsup: from me

---

Reviewed 3 of 3 files at r5.
Review status: all files reviewed at latest revision, 3 unresolved discussions, some commit checks failed.

---



Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/uproxy/uproxy/1754)
<!-- Sent from Reviewable.io -->

kennysong commented:
Fixed a merge error with the new variable names for i18n. Does this look good @jpevarnek?

jpevarnek commented:
Still looks fine to me :)

---

Reviewed 3 of 3 files at r6.
Review status: all files reviewed at latest revision, 3 unresolved discussions, some commit checks failed.

---



Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/uproxy/uproxy/1754)
<!-- Sent from Reviewable.io -->
